File: Typecast/TCResourceMap.m
//
//  TCResourceMap.m
//  Type Designer
//
//  Created by David Schweinsberg on 28/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCResourceMap.h"
#import "TCDataInput.h"
#import "TCResourceType.h"

@implementation TCResourceMap

- (id)initWithDataInput:(TCDataInput *)dataInput
{
    self = [super init];
    if (self)
    {
        _headerData = [dataInput readDataWithLength:16];
        _nextResourceMap = [dataInput readInt];
        _fileReferenceNumber = [dataInput readUnsignedShort];
        _attributes = [dataInput readUnsignedShort];

        int typeOffset = [dataInput readUnsignedShort];
        int nameOffset = [dataInput readUnsignedShort];
        int typeCount = [dataInput readUnsignedShort] + 1;

        // Read types
        NSMutableArray *types = [[NSMutableArray alloc] initWithCapacity:typeCount];
        for (int i = 0; i < typeCount; ++i)
            [types addObject:[[TCResourceType alloc] initWithDataInput:dataInput]];
        _types = types;

        // Read the references
        for (int i = 0; i < typeCount; ++i)
            [_types[i] readReferences:dataInput];

        // Read the names
        for (int i = 0; i < typeCount; ++i)
            [_types[i] readNames:dataInput];
    }
    return self;
}

- (TCResourceType *)resourceTypeWithName:(NSString *)typeName
{
    for (TCResourceType *type in _types)
    {
        NSString *s = [type typeAsString];
        if ([s isEqualToString:typeName])
            return type;
    }
    return nil;
}

@end

File: Typecast/TCResourceMap.swift
//
//  TCResourceMap.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/27/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation

class TCResourceMap {
  let headerData: Data
  let nextResourceMap: Int
  let fileReferenceNumber: Int
  let attributes: Int
  var types: [TCResourceType]

  init(dataInput: TCDataInput) {
    headerData = dataInput.readData(withLength:16)
    nextResourceMap = Int(dataInput.readInt())
    fileReferenceNumber = Int(dataInput.readUnsignedShort())
    attributes = Int(dataInput.readUnsignedShort())

    _ = dataInput.readUnsignedShort()  // typeOffset
    _ = dataInput.readUnsignedShort()  // nameOffset
    let typeCount = Int(dataInput.readUnsignedShort()) + 1

    // Read types
    types = []
    for _ in 0 ..< typeCount {
      types.append(TCResourceType(dataInput: dataInput))
    }

    // Read the references
    for i in 0 ..< typeCount {
      types[i].readReferences(dataInput: dataInput)
    }

    // Read the names
    for i in 0 ..< typeCount {
      types[i].readNames(dataInput: dataInput)
    }
  }

  func resourceType(name typeName: String) -> TCResourceType? {
    for type in types {
      if type.typeAsString() == typeName {
        return type;
      }
    }
    return nil;
  }
}
